DevTools: Add Movement Flags DevTools Plugin.

Pratted opened the pull request:
Background: I created this tool to assist a Hallowed Sepulchre plugin (in progress) and I thought it might be useful to others.

This plugin highlights the tile the player is currently hovering over and displays the movement flags above it. The movement flags can be useful for pathfinding and other debugging purposes.

<img width="762" alt="Screen Shot 2020-08-26 at 2 03 54 PM" src="https://user-images.githubusercontent.com/19481676/91341554-a9a5ee00-e7a7-11ea-9c15-7094b48fed2f.png">
<img width="762" alt="Screen Shot 2020-08-26 at 2 04 07 PM" src="https://user-images.githubusercontent.com/19481676/91341546-a7439400-e7a7-11ea-9734-4d82c19c9e07.png">
<img width="762" alt="Screen Shot 2020-08-26 at 2 04 26 PM" src="https://user-images.githubusercontent.com/19481676/91341522-9e52c280-e7a7-11ea-8f2d-c372dede0865.png">
<img width="762" alt="Screen Shot 2020-08-26 at 2 04 36 PM" src="https://user-images.githubusercontent.com/19481676/91341507-9b57d200-e7a7-11ea-8933-a936ef0cd153.png">

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/DevToolsOverlay.java:
```diff
@@ -249,6 +255,34 @@ private void renderTileTooltip(Graphics2D graphics, Tile tile)
+
+				if (movementFlags.isEmpty())
+				{
+					toolTipManager.add(new Tooltip("No flags"));
```
```suggestion
					toolTipManager.add(new Tooltip("No movement flags"));
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/MovementFlag.java:
```diff
@@ -0,0 +1,70 @@
+	BLOCK_MOVEMENT_FULL(CollisionDataFlag.BLOCK_MOVEMENT_FULL);
+
+	@Getter
+	int flag;
```
```suggestion
	private int flag;
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/MovementFlag.java:
```diff
@@ -0,0 +1,70 @@
+	public static Set<MovementFlag> getSetFlags(int collisionData)
+	{
+		return Arrays.stream(values())
+				.filter(movementFlag -> (movementFlag.flag & collisionData) > 0)
+				.collect(Collectors.toSet());
```
Dedent these lines by one level.
```suggestion
			.filter(movementFlag -> (movementFlag.flag & collisionData) > 0)
			.collect(Collectors.toSet());
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/DevToolsOverlay.java:
```diff
@@ -249,6 +255,34 @@ private void renderTileTooltip(Graphics2D graphics, Tile tile)
+
+		if (poly != null && poly.contains(client.getMouseCanvasPosition().getX(), client.getMouseCanvasPosition().getY()))
+		{
+			// We can only get CollisionMaps only for instanced regions.
```
I believe we have reversed this behavior, so this should be available in non-instanced areas now.

Pratted replied:
Ok. I removed the isInInstancedRegion() check below and tested it in a few non-instanced regions. It worked in all of the regions I tested so i'll keep that way.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/DevToolsOverlay.java:
```diff
@@ -249,6 +255,34 @@ private void renderTileTooltip(Graphics2D graphics, Tile tile)
+	{
+		Polygon poly = Perspective.getCanvasTilePoly(client, tile.getLocalLocation());
+
+		if (poly != null && poly.contains(client.getMouseCanvasPosition().getX(), client.getMouseCanvasPosition().getY()))
```
Use a guard clause instead of a large nested if statement.

eg.

```java
if (poly == null || !poly.contains(client.getMousePosition().getX(), client.getMousePosition().getY())
{
	return;
}

// continue method code here
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/DevToolsOverlay.java:
```diff
@@ -249,6 +255,34 @@ private void renderTileTooltip(Graphics2D graphics, Tile tile)
+			// We can only get CollisionMaps only for instanced regions.
+			if (client.isInInstancedRegion() && client.getCollisionMaps() != null)
+			{
+				int flags[][] = client.getCollisionMaps()[client.getPlane()].getFlags();
```
We prefer `Type[][] varName` over `Type varName[][]`

Pratted commented:
@Nightfirecat FYI, I addressed your feedback on this PR. I marked all all conversations as resolved and pushed the appropriate changes - I'm not sure why it still says "Changes Requested".

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/MovementFlag.java:
```diff
@@ -0,0 +1,70 @@
+ * @see CollisionDataFlag
+ */
+@AllArgsConstructor
+public enum MovementFlag
```
This should be package private
```suggestion
enum MovementFlag
```

Nightfirecat reviewed (commented):
Generally LGTM now

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/devtools/MovementFlag.java:
```diff
@@ -0,0 +1,70 @@
+	public static Set<MovementFlag> getSetFlags(int collisionData)
+	{
+		return Arrays.stream(values())
+			.filter(movementFlag -> (movementFlag.flag & collisionData) > 0)
```
This does work, but note if a flag was `0x80000000` it would not work due to ints in Java being signed. Normally you would use !=
